File: ios/FirebaseTOTPModule.swift
import ExpoModulesCore
import FirebaseAuth
import FirebaseCore

public class FirebaseTOTPModule: Module {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('FirebaseTOTP')` in JavaScript.
    Name("FirebaseTOTP")

    // Defines event names that the module can send to JavaScript.
    Events("onEnrollmentComplete", "onVerificationComplete", "onError")

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("enrollUserInTOTP") { (userId: String?, accountName: String?, issuer: String?) in
      do {
        let user = try self.getFirebaseUser(userId)
        
        // Start the TOTP enrollment process
        let multiFactorSession = try await user.multiFactor.session()
        
        // Create a TOTP enrollment
        let totpSecret = try await TOTPMultiFactorGenerator.generateSecret(with: multiFactorSession)
        
        // Get the shared secret
        let sharedSecretKey = totpSecret.sharedSecretKey
        
        // Use provided parameters or defaults
        let finalAccountName = accountName ?? user.email ?? "default_account"
        let finalIssuer = issuer ?? "FirebaseTOTP"
        
        // Generate QR code URL
        let qrCodeURL = totpSecret.generateQRCodeURL(
            withAccountName: finalAccountName,
            issuer: finalIssuer
        )
        
        // Store the TOTP secret for later verification
        // In a real app, you might want to store this in a more secure way
        self.totpSecretForVerification = totpSecret
        
        // Return the enrollment information
        return [
          "secretKey": sharedSecretKey,
          "qrCodeUrl": qrCodeURL
        ]
      } catch {
        self.sendEvent("onError", [
          "error": error.localizedDescription
        ])
        throw error
      }
    }
    
    AsyncFunction("verifyTOTPCode") { (code: String, userId: String?) in
      do {
        let user = try self.getFirebaseUser(userId)
        
        // Make sure we have a TOTP secret to verify against
        guard let totpSecret = self.totpSecretForVerification else {
          throw NSError(domain: "FirebaseTOTP", code: 4, userInfo: [NSLocalizedDescriptionKey: "No TOTP secret available for verification. Please enroll first."])
        }
        
        // Create the assertion for enrollment
        let multiFactorAssertion = TOTPMultiFactorGenerator.assertionForEnrollment(
          with: totpSecret,
          oneTimePassword: code
        )
        
        // Enroll the user with the credential
        try await user.multiFactor.enroll(
          with: multiFactorAssertion,
          displayName: "TOTP"
        )
        
        // Clear the stored secret after successful verification
        self.totpSecretForVerification = nil
        
        // Return success
        return [
          "success": true,
          "message": "TOTP verification successful"
        ]
      } catch {
        self.sendEvent("onError", [
          "error": error.localizedDescription
        ])
        
        return [
          "success": false,
          "message": error.localizedDescription
        ]
      }
    }
    
    AsyncFunction("disableTOTP") { (userId: String?) in
      do {
        let user = try self.getFirebaseUser(userId)
        
        // Get the enrolled factors
        let enrolledFactors = user.multiFactor.enrolledFactors
        
        // Find the TOTP factor
        if let totpFactor = enrolledFactors.first(where: { $0.factorID == TOTPMultiFactorGenerator.factorID }) {
          // Unenroll the TOTP factor
          try await user.multiFactor.unenroll(with: totpFactor)
        }
      } catch {
        self.sendEvent("onError", [
          "error": error.localizedDescription
        ])
        throw error
      }
    }
  }
  
  // Store the TOTP secret between enrollment and verification
  private var totpSecretForVerification: TOTPSecret?
  
  // Helper method to get the Firebase user
  private func getFirebaseUser(_ userId: String?) throws -> User {
    let auth = Auth.auth()
    
    if let userId = userId, !userId.isEmpty {
      // In iOS, we can't directly get a user by ID from the client SDK
      // We should use the current user and verify the ID matches
      guard let currentUser = auth.currentUser, currentUser.uid == userId else {
        throw NSError(domain: "FirebaseTOTP", code: 2, userInfo: [NSLocalizedDescriptionKey: "User not found or ID doesn't match current user"])
      }
      return currentUser
    } else {
      // Get the current user
      guard let user = auth.currentUser else {
        throw NSError(domain: "FirebaseTOTP", code: 3, userInfo: [NSLocalizedDescriptionKey: "No user is currently signed in"])
      }
      return user
    }
  }
}

File: android/src/main/java/com/uluru/fb/totp/FirebaseTOTPModule.kt
package com.uluru.fb.totp

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.MultiFactor
import com.google.firebase.auth.PhoneMultiFactorGenerator
import com.google.firebase.auth.PhoneMultiFactorInfo
import kotlinx.coroutines.tasks.await

class FirebaseTOTPModule : Module() {
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  override fun definition() = ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('FirebaseTOTP')` in JavaScript.
    Name("FirebaseTOTP")

    // Defines event names that the module can send to JavaScript.
    Events("onEnrollmentComplete", "onVerificationComplete", "onError")

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
    AsyncFunction("enrollUserInTOTP") { userId: String?, accountName: String?, issuer: String? ->
      try {
        val user = getFirebaseUser(userId)
        
        // Note: Firebase Android SDK doesn't directly support TOTP yet
        // This is a placeholder implementation
        // In a real implementation, you would use a custom TOTP library
        
        // Generate a random secret key (base32 encoded)
        val secretKey = generateTOTPSecret()
        
        // Generate a verification ID (this would be used to link the TOTP with the user)
        val verificationId = java.util.UUID.randomUUID().toString()
        
        // Use provided parameters or defaults
        val finalAccountName = accountName ?: user.email ?: "user"
        val finalIssuer = issuer ?: "FirebaseTOTP"
        
        // Generate a QR code URL (using standard TOTP URI format)
        val qrCodeUrl = generateTOTPQRCodeUrl(finalAccountName, finalIssuer, secretKey)
        
        // Return the enrollment information
        mapOf(
          "secretKey" to secretKey,
          "qrCodeUrl" to qrCodeUrl,
          "verificationId" to verificationId
        )
      } catch (e: Exception) {
        sendEvent("onError", mapOf("error" to e.localizedMessage))
        throw e
      }
    }
    
    AsyncFunction("verifyTOTPCode") { code: String, verificationId: String, userId: String? ->
      try {
        // Note: Firebase Android SDK doesn't directly support TOTP yet
        // This is a placeholder implementation
        // In a real implementation, you would verify the TOTP code against the secret
        
        // For demonstration, we'll just check if the code is 6 digits
        val isValid = code.length == 6 && code.all { it.isDigit() }
        
        if (isValid) {
          mapOf(
            "success" to true,
            "message" to "TOTP verification successful"
          )
        } else {
          mapOf(
            "success" to false,
            "message" to "Invalid TOTP code"
          )
        }
      } catch (e: Exception) {
        sendEvent("onError", mapOf("error" to e.localizedMessage))
        mapOf(
          "success" to false,
          "message" to e.localizedMessage
        )
      }
    }
    
    AsyncFunction("disableTOTP") { userId: String? ->
      try {
        // Note: Firebase Android SDK doesn't directly support TOTP yet
        // This is a placeholder implementation
        // In a real implementation, you would remove the TOTP from the user's account
        
        // For demonstration, we'll just return success
        Unit
      } catch (e: Exception) {
        sendEvent("onError", mapOf("error" to e.localizedMessage))
        throw e
      }
    }
  }
  
  // Helper method to get the Firebase user
  private fun getFirebaseUser(userId: String?): FirebaseUser {
    val auth = FirebaseAuth.getInstance()
    
    if (!userId.isNullOrEmpty()) {
      // Note: Firebase Android SDK doesn't have a direct way to get a user by ID
      // In a real implementation, you would need to use Firebase Admin SDK or a custom backend
      throw Exception("Getting user by ID is not supported in the Android client SDK")
    } else {
      // Get the current user
      return auth.currentUser ?: throw Exception("No user is currently signed in")
    }
  }
  
  // Helper method to generate a random TOTP secret
  private fun generateTOTPSecret(): String {
    val chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567" // Base32 character set
    val random = java.security.SecureRandom()
    val secretLength = 16 // 16 characters = 80 bits
    
    return (1..secretLength)
      .map { chars[random.nextInt(chars.length)] }
      .joinToString("")
  }
  
  // Helper method to generate a TOTP QR code URL
  private fun generateTOTPQRCodeUrl(username: String, issuer: String, secret: String): String {
    val encodedIssuer = java.net.URLEncoder.encode(issuer, "UTF-8")
    val encodedUsername = java.net.URLEncoder.encode(username, "UTF-8")
    
    return "otpauth://totp/$encodedIssuer:$encodedUsername?secret=$secret&issuer=$encodedIssuer&algorithm=SHA1&digits=6&period=30"
  }
}
